### PR TITLE
fix: Add error message for duplicate route param name

### LINKED ISSUE #7989: Next.js does not error when dynamic route segments re-use a name
Next.js does not throw an error for `pages/[id]/[id].js`, as reported [on Spectrum](https://spectrum.chat/next-js/general/nextpagecontext-query-is-wrongly-typed-as-string-string~3baa1635-9aa8-4531-8f6a-79de8698d827).

### PR DESCRIPTION
**Pull request by nataliemarleny**
This PR adds an error message for when there is a duplicate slug name within a dynamic path.

Currently WIP - going to add an integration test for changes made in 7240214

fixes: https://github.com/zeit/next.js/issues/7989

### CODE REVIEW

File: packages/next-server/lib/router/utils/sorted-routes.ts [hunk 1]
```diff
@@ -2,6 +2,11 @@ class UrlNode {
   placeholder: boolean = true
   children: Map<string, UrlNode> = new Map()
   slugName: string | null = null
+  parentSlugNames: Array<string> = new Array()
```
Comment (timneutkens) [inline packages/next-server/lib/router/utils/sorted-routes.ts]: Shouldn't this be:

```suggestion
  parentSlugNames: Array<string> = []
```

File: packages/next-server/lib/router/utils/sorted-routes.ts [hunk 2]
```diff
@@ -2,6 +2,11 @@ class UrlNode {
...
+
+  constructor(parentSlugNames = new Array()) {
```
Comment (timneutkens) [inline packages/next-server/lib/router/utils/sorted-routes.ts]: Same as above:

```suggestion
  constructor(parentSlugNames = []) {
```

File: packages/next-server/lib/router/utils/sorted-routes.ts [hunk 3]
```diff
@@ -53,12 +58,19 @@ class UrlNode {
         )
       }
 
+      if (this.parentSlugNames.indexOf(slugName) !== -1) {
```
Comment (timneutkens) [inline packages/next-server/lib/router/utils/sorted-routes.ts]: Should this be in development only 🤔

Comment (nataliemarleny) [reply packages/next-server/lib/router/utils/sorted-routes.ts]: I'm not following - please could you clarify? 

Did you mean - should this error be triggered only in development?

Comment (timneutkens) [reply packages/next-server/lib/router/utils/sorted-routes.ts]: > Did you mean - should this error be triggered only in development?

Yep! I'm not entirely sure though!

File: packages/next-server/lib/router/utils/sorted-routes.ts (hunk 1 above)
Comment (timneutkens) [reply packages/next-server/lib/router/utils/sorted-routes.ts]: You might want to use `new Set` here though, as everything has to be unique.

**Review by Timer — COMMENTED** [review]
Can we please add this case to the unit tests?

https://github.com/zeit/next.js/blob/canary/test/unit/page-route-sorter.test.js

**Comment by Timer** [comment]
@nataliemarleny Awesome work updating this code for clarity 🙏!

I've went ahead and added the unit test so we can land this and start having users / contributors benefit from your changes.

Thanks much!

**Review by Timer — APPROVED** [review]

**Review by ijjk — APPROVED** [review]

**Pull request merged**